lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -1114,44 +1145,74 @@ def _get_app_or_unit(self) -> Union[Application, Unit]:
     @property
     def private_key(self) -> Optional[PrivateKey]:
         """Return the private key."""
+        if self._private_key:
+            return self._private_key
         if not self._private_key_generated():
             return None
         secret = self.charm.model.get_secret(label=self._get_private_key_secret_label())
         private_key = secret.get_content(refresh=True)["private-key"]
         return PrivateKey.from_string(private_key)
 
-    def _generate_private_key(self) -> None:
+    def _ensure_private_key(self) -> None:
+        if self._private_key:
+            self._remove_private_key_secret()
+            return
         if self._private_key_generated():
+            logger.debug("Private key already generated")
             return
-        private_key = generate_private_key()
-        self.charm.unit.add_secret(
-            content={"private-key": str(private_key)},
-            label=self._get_private_key_secret_label(),
-        )
-        logger.info("Private key generated")
+        self._generate_private_key()
 
     def regenerate_private_key(self) -> None:
         """Regenerate the private key.
 
         Generate a new private key, remove old certificate requests and send new ones.
+
+        Raises:
+            TLSCertificatesError: If the private key is managed by the charm.
+                In that case the charm is responsible for regenerating the private key
+                and the only entry point is the init function.
         """
+        if self._private_key:
+            raise TLSCertificatesError(
+                "Private key managed by the charm, this function can't be used"
```

Comment:
I'm not sure about the phrasing "managed by the charm". I would prefer "provided through the private_key parameter"  or something like that

Reply:
Done.